File: fatirlar-play-app/app/service/AdvertService.java
package service;

import model.Advert;
import model.District;
import model.rest.SearchRequest;
import model.ui.AdvertPrices;
import repository.AdvertRepository;

import javax.inject.Inject;
import javax.inject.Singleton;

import java.util.*;

@Singleton
public class AdvertService {

    private static final int ADVERTS_PER_REQUEST = 15;
    private static final int ADVERTS_ON_MAIN_PAGE = 9;

    @Inject
    private AdvertRepository advertMapper;

    public List<Advert> getAdvertsForMainPage() {
        return advertMapper.getNextAdvertsBeforeTime(System.currentTimeMillis(), ADVERTS_ON_MAIN_PAGE);
    }

    public int getPagesCount(SearchRequest searchRequest) {
        Objects.requireNonNull(searchRequest.getPriceRange());

        if (searchRequest.getPriceRange().size() != 2) {
            throw new IllegalStateException("Incorrect price range");
        }

        Collection<District> districts = searchRequest.getDistricts();
        if (districts == null || districts.isEmpty()) {
            districts = EnumSet.allOf(District.class);
        }

        int priceFrom = searchRequest.getPriceRange().get(0) * 1000;
        int priceTo = searchRequest.getPriceRange().get(1) * 1000;

        List<Integer> rooms = new ArrayList<>();
        if (searchRequest.isRoom1()) {
            rooms.add(1);
        }
        if (searchRequest.isRoom2()) {
            rooms.add(2);
        }
        if (searchRequest.isRoom3()) {
            rooms.add(3);
        }

        int advertsCount = advertMapper.getAdvertsCount(districts, priceFrom, priceTo, rooms);
        return advertsCount / ADVERTS_PER_REQUEST;
    }

    public List<Advert> getAdverts(SearchRequest searchRequest) {
        Objects.requireNonNull(searchRequest.getPriceRange());

        if (searchRequest.getPage() < 0) {
            throw new IllegalArgumentException("Illegal page number");
        }

        if (searchRequest.getPriceRange().size() != 2) {
            throw new IllegalStateException("Incorrect price range");
        }

        Collection<District> districts = searchRequest.getDistricts();
        if (districts == null || districts.isEmpty()) {
            districts = EnumSet.allOf(District.class);
        }

        int priceFrom = searchRequest.getPriceRange().get(0) * 1000;
        int priceTo = searchRequest.getPriceRange().get(1) * 1000;

        List<Integer> rooms = new ArrayList<>();
        if (searchRequest.isRoom1()) {
            rooms.add(1);
        }
        if (searchRequest.isRoom2()) {
            rooms.add(2);
        }
        if (searchRequest.isRoom3()) {
            rooms.add(3);
        }

        return advertMapper.getAdverts(districts, priceFrom, priceTo, rooms, searchRequest.getPage() * ADVERTS_PER_REQUEST, ADVERTS_PER_REQUEST);
    }

    public Advert getAdvert(long id) {
        return advertMapper.findById(id);
    }

    public AdvertPrices getAdvertPrices() {
        return advertMapper.getAdvertPrices();
    }

    public void removeAdvert(long advertId) {
        advertMapper.deleteAdvert(advertId);
    }

}

File: fatirlar-play-app/app/service/AdvertService.scala
package service

import java.util
import javax.inject.{Inject, Singleton}

import model.District
import model.rest.SearchParameters
import repository.interops.AdvertRepositoryJv

import scala.collection.JavaConverters._

@Singleton
class AdvertService @Inject()(advertRepositoryJv: AdvertRepositoryJv) {

  val advertsOnMainPage = 9
  val advertsPerRequest = 15

  def advertsForMainPage = advertRepositoryJv.getNextAdvertsBeforeTime(System.currentTimeMillis, advertsOnMainPage)

  def pagesCount(searchParameters: SearchParameters): Int = {
    val searchParameters = prepareSearchParameters(searchParameters)
    advertRepositoryJv.getAdvertsCount(searchParameters._1.asJavaCollection,
      searchParameters._3._1, searchParameters._3._2,
      searchParameters._2.map(v => v.asInstanceOf[Integer]).asJava) / advertsPerRequest
  }

  def adverts(searchParameters: SearchParameters) = {
    val searchParameters = prepareSearchParameters(searchParameters)
    advertRepositoryJv.getAdverts(
      searchParameters._1.asJavaCollection,
      searchParameters._3._1, searchParameters._3._2,
      searchParameters._2.map(v => v.asInstanceOf[Integer]).asJava,
      searchParameters._4.getOrElse(0) * advertsPerRequest,
      advertsPerRequest
    )
  }

  def advert(id: Int) = advertRepositoryJv.findById(id)

  private def prepareSearchParameters(searchParameters: SearchParameters) = (
    if (searchParameters.districts.isEmpty) util.EnumSet.allOf(classOf[District]).asScala.toList else searchParameters.districts,
    List(roomsVal(searchParameters.rooms1, 1), roomsVal(searchParameters.rooms2, 2), roomsVal(searchParameters.rooms3, 3)).flatten,
    (searchParameters.priceRange._1 * 1000, searchParameters.priceRange._2 * 1000),
    searchParameters.page
  )

  private def roomsVal(has: Boolean, room: Int): Option[Int] = if (has) Some(room) else None

}
